### PR TITLE
Add Challenges Support to LBP3

### PR DESCRIPTION
**Pull request by W0lf4llo**
Please let me know if there's better ways of doing this, thanks! :D
image

### CODE REVIEW

File: ProjectLighthouse/ChallengeConfig.xml
Comment (sudokoko) [inline ProjectLighthouse/ChallengeConfig.xml:1]: This file could be represented as a class/entity, and then served out by manually building the XML using `LbpCustomXml`, seeing as the format of this is a little weird. I wouldn't mind helping with this once I get home.

**Review by sudokoko — CHANGES_REQUESTED** [review]

File: ProjectLighthouse/ChallengeConfig.xml
Comment (W0lf4llo) [reply ProjectLighthouse/ChallengeConfig.xml:1]: Alright, thank you :D
Just give me a call whenever you're free!

**Review by sudokoko — APPROVED** [review]

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+public class GameChallenge : ILbpSerializable
+{
+    [XmlElement("Total_challenges")]
+    public int TotalChallenges { get; set; }
+
+    [XmlElement("Challenge_End_Date")]
+    public ulong EndDate { get; set; }
+
+    [XmlElement("Challenge_Top_Rank_Bronze_Range")]
+    public decimal TopRankBronzePercentage { get; set; }
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:16]:
```suggestion
    public float BronzeRankPercentage { get; set; }
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+    [XmlElement("Challenge_Top_Rank_Silver_Range")]
+    public decimal TopRankSilverPercentage { get; set; }
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:19]:
```suggestion
    public float SilverRankPercentage { get; set; }
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+    [XmlElement("Challenge_Top_Rank_Gold_Range")]
+    public decimal TopRankGoldPercentage { get; set; }
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:22]:
```suggestion
    public float GoldRankPercentage { get; set; }
```
These don't need to be decimals if they only represent percentages

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+    [XmlElement("Challenge_CycleTime")]
+    public ulong CycleTime { get; set; }
+
+    // ReSharper disable once IdentifierTypo
+    [XmlElement("item_data")]
+    public List<ChallengeItemData> ChallengeItemDatas { get; set; }
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:29]:
```suggestion
    public List<GameChallenge> Challenges { get; set; }
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    #endregion
+}
+
+public class ChallengeItemData
+{
+    [XmlAttribute("Challenge_ID")]
+    public int Id { get; set; }
+
+    [XmlAttribute("Challenge_active_date_starts")]
+    public ulong ActiveDateStarts { get; set; }
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:254]:
```suggestion
    public ulong StartTime { get; set; }
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+    [XmlAttribute("Challenge_active_date_ends")]
+    public ulong ActiveDateEnds { get; set; }
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:257]:
```suggestion
    public ulong EndTime { get; set; }
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    public string ContentName { get; set; }
+
+    [XmlAttribute("Challenge_Planet_User")]
+    public string PlanetUser { get; set; }
+
+    [XmlAttribute("Challenge_planetId")]
+    public ulong PlanetId { get; set; }
+
+    [XmlAttribute("Challenge_photo_1")]
+    public ulong Photo1 { get; set; }
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:284]:
```suggestion
    public ulong PhotoId { get; set; }
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 8]
```diff
@@ -0,0 +1,285 @@
+﻿using System.Collections.Generic;
+using System.Xml.Serialization;
+
+namespace LBPUnion.ProjectLighthouse.Types.Serialization;
+
+[XmlRoot("Challenge_header")]
+public class GameChallenge : ILbpSerializable
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:7]:
```suggestion
public partial class GameChallengeResponse : ILbpSerializable
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+                PlanetId = 1085254,
+                Photo1 = 1112663,
+            },
+        },
+    };
+
+    #endregion
+}
+
+public class ChallengeItemData
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:248]:
```suggestion
public class GameChallenge
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+namespace LBPUnion.ProjectLighthouse.Types.Serialization;
+
+[XmlRoot("Challenge_header")]
+public class GameChallenge : ILbpSerializable
+{
+    [XmlElement("Total_challenges")]
+    public int TotalChallenges { get; set; }
+
+    [XmlElement("Challenge_End_Date")]
+    public ulong EndDate { get; set; }
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:13]:
```suggestion
    public ulong EndTime { get; set; }
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+                RankPin = 3721717765,
+                ContentName = "TG_LittleBigPlanet3",
+                PlanetUser = "qd3c781a5a6-GBen",
+                PlanetId = 1085254,
+                Photo1 = 1112663,
+            },
+        },
+    };
+
+    #endregion
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:245]:
```suggestion
```

File: ProjectLighthouse/Types/Serialization/GameChallenge.cs [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+
+    [XmlAttribute("Challenge_Planet_User")]
+    public string PlanetUser { get; set; }
+
+    [XmlAttribute("Challenge_planetId")]
+    public ulong PlanetId { get; set; }
+
+    [XmlAttribute("Challenge_photo_1")]
+    public ulong Photo1 { get; set; }
+}
```
Comment (Slendy) [inline ProjectLighthouse/Types/Serialization/GameChallenge.cs:285]: I'd rather the hardcoded definition be at the bottom of the file and the actual serializable classes be at the top so you can split it into a partial and then put the ServerChallenges at the bottom
```suggestion
}

public partial class GameChallengeResponse {
    public static GameChallengeResponse ServerChallenges() => new()
    {
        TotalChallenges = 14,
        EndTime = 1494460860000000,
        BronzeRankPercentage = 0.51f,
        SilverRankPercentage = 0.26f,
        GoldRankPercentage = 0.11f,
        CycleTime = 1209600000000,
        Challenges = new List<GameChallenge>
        {
            new()
            {
                Id = 0,
                StartTime = 1476096166000000,
                EndTime = 1478736060000000,
                LamsDescriptionId = "CHALLENGE_NEWTONBOUNCE_DESC",
                LamsTitleId = "CHALLENGE_NEWTONBOUNCE_NAME",
                PinId = 3003874881,
                RankPin = 2922567456,
                ContentName = "TG_LittleBigPlanet3",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1085260,
                PhotoId = 1112639,
            },
            new()
            {
                Id = 1,
                StartTime = 1478736060000000,
                EndTime = 1479945660000000,
                LamsDescriptionId = "CHALLENGE_SCREENCHASE_DESC",
                LamsTitleId = "CHALLENGE_SCREENCHASE_NAME",
                PinId = 282407472,
                RankPin = 3340696069,
                ContentName = "TG_LittleBigPlanet2",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1102387,
                PhotoId= 1112651,
            },
            new()
            {
                Id = 2,
                StartTime = 1479945660000000,
                EndTime = 1481155260000000,
                LamsDescriptionId = "CHALLENGE_RABBITBOXING_DESC",
                LamsTitleId = "CHALLENGE_RABBITBOXING_NAME",
                PinId = 2529088759,
                RankPin = 958144818,
                ContentName = "TG_LittleBigPlanet2",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1085264,
                PhotoId = 1112627,
            },
            new()
            {
                Id = 3,
                StartTime = 1481155260000000,
                EndTime = 1481414460000000,
                LamsDescriptionId = "CHALLENGE_FLOATYFLUID_DESC",
                LamsTitleId = "CHALLENGE_FLOATYFLUID_NAME",
                PinId = 183892581,
                RankPin = 3442917932,
                Content = "LBPDLCNISBLK0001",
                ContentName = "SBSP_THEME_PACK_NAME",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1095449,
                PhotoId = 1112619,
            },
            new()
            {
                Id = 4,
                StartTime = 1481414460000000,
                EndTime = 1483574460000000,
                LamsDescriptionId = "CHALLENGE_ISLANDRACE_DESC",
                LamsTitleId = "CHALLENGE_ISLANDRACE_NAME",
                PinId = 315245769,
                RankPin = 443310584,
                ContentName = "TG_LittleBigPlanet",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1102858,
                PhotoId = 1112655,
            },
            new()
            {
                Id = 5,
                StartTime = 1483574460000000,
                EndTime = 1484784060000000,
                LamsDescriptionId = "CHALLENGE_SPACEDODGING_DESC",
                LamsTitleId = "CHALLENGE_SPACEDODGING_NAME",
                PinId = 144212050,
                RankPin = 2123417147,
                ContentName = "TG_LittleBigPlanet3",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1085266,
                PhotoId = 1112667,
            },
            new()
            {
                Id = 6,
                StartTime = 1484784060000000,
                EndTime = 1485993660000000,
                LamsDescriptionId = "CHALLENGE_INVISIBLECIRCUIT_DESC",
                LamsTitleId = "CHALLENGE_INVISIBLECIRCUIT_NAME",
                PinId = 249569175,
                RankPin = 1943114258,
                ContentName = "TG_LittleBigPlanet",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1096814,
                PhotoId = 1112635,
            },
            new()
            {
                Id = 7,
                StartTime = 1485993660000000,
                EndTime = 1487203260000000,
                LamsDescriptionId = "CHALLENGE_HOVERBOARDRAILS_DESC",
                LamsTitleId = "CHALLENGE_HOVERBOARDRAILS_NAME",
                PinId = 3478661003,
                RankPin = 592022798,
                Content = "LBPDLCBTTFLK0001",
                ContentName = "BTTF_LEVEL_KIT_NAME",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1085256,
                PhotoId = 1112623,
            },
            new()
            {
                Id = 8,
                StartTime = 1487203260000000,
                EndTime = 1488412860000000,
                LamsDescriptionId = "CHALLENGE_TOWERBOOST_DESC",
                LamsTitleId = "CHALLENGE_TOWERBOOST_NAME",
                PinId = 216730878,
                RankPin = 545532447,
                ContentName = "TG_LittleBigPlanet2",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1092504,
                PhotoId = 1112671,
            },
            new()
            {
                Id = 9,
                StartTime = 1488412860000000,
                EndTime = 1489622460000000,
                LamsDescriptionId = "CHALLENGE_SWOOPPANELS_DESC",
                LamsTitleId = "CHALLENGE_SWOOPPANELS_NAME",
                PinId = 2054302637,
                RankPin = 3288689476,
                ContentName = "TG_LittleBigPlanet2",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1085268,
                PhotoId = 1112643,
            },
            new()
            {
                Id = 10,
                StartTime = 1489622460000000,
                EndTime = 1490832060000000,
                LamsDescriptionId = "CHALLENGE_PINBALLCRYPTS_DESC",
                LamsTitleId = "CHALLENGE_PINBALLCRYPTS_NAME",
                PinId = 618998172,
                RankPin = 4087839785,
                ContentName = "TG_LittleBigPlanet3",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1085262,
                PhotoId = 1112647,
            },
            new()
            {
                Id = 11,
                StartTime = 1490832060000000,
                EndTime = 1492041660000000,
                LamsDescriptionId = "CHALLENGE_TIEHOP_DESC",
                LamsTitleId = "CHALLENGE_TIEHOP_NAME",
                PinId = 3953447125,
                RankPin = 2556445436,
                ContentName = "TG_LittleBigPlanet",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1092367,
                PhotoId = 1112659,
            },
            new()
            {
                Id = 12,
                StartTime = 1492041660000000,
                EndTime = 1493251260000000,
                LamsDescriptionId = "CHALLENGE_JOKERFUNHOUSE_DESC",
                LamsTitleId = "CHALLENGE_JOKERFUNHOUSE_NAME",
                PinId = 1093784294,
                RankPin = 1757295127,
                Content = "LBPDLCWBDCLK0001",
                ContentName = "DCCOMICS_THEME_NAME",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1085258,
                PhotoId = 1112631,
            },
            new()
            {
                Id = 13,
                StartTime = 1493251260000000,
                EndTime = 1494460860000000,
                LamsDescriptionId = "CHALLENGE_DINERSHOOTING_DESC",
                LamsTitleId = "CHALLENGE_DINERSHOOTING_NAME",
                PinId = 1568570416,
                RankPin = 3721717765,
                ContentName = "TG_LittleBigPlanet3",
                PlanetUser = "qd3c781a5a6-GBen",
                PlanetId = 1085254,
                PhotoId = 1112663,
            },
        },
    };
}
```

**Review by Slendy — CHANGES_REQUESTED** [review]
Just a few naming nitpicks

**Comment by sudokoko** [comment]
My dearest condolences to you and your family in this trying time.

File: ProjectLighthouse.Servers.GameServer/Controllers/Login/ClientConfigurationController.cs [hunk 13]
```diff
@@ -1,8 +1,10 @@
 #nullable enable
 using System.Diagnostics.CodeAnalysis;
+using System.Text;
 using LBPUnion.ProjectLighthouse.Configuration;
 using LBPUnion.ProjectLighthouse.Database;
 using LBPUnion.ProjectLighthouse.Extensions;
+using LBPUnion.ProjectLighthouse.Serialization;
```
Comment (Zaprit) [inline ProjectLighthouse.Servers.GameServer/Controllers/Login/ClientConfigurationController.cs:7]: Remove unused imports
```suggestion
using LBPUnion.ProjectLighthouse.Configuration;
using LBPUnion.ProjectLighthouse.Database;
using LBPUnion.ProjectLighthouse.Extensions;
```

**Review by Zaprit — CHANGES_REQUESTED** [review]
See changes

**Review by Zaprit — APPROVED** [review]

**Review by Slendy — APPROVED** [review]

**Pull request merged**